### PR TITLE
Add Psalm integration

### LINKED ISSUE #26: Psalm integration
### Feature Request

|    Q        |   A
|------------ | ------
| QA    | yes

#### Summary

As decided during the [Technical-Steering-Committee Meeting on August 3rd, 2020](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-static-analysis-tooling), Laminas wants to implement [vimeo/psalm](https://github.com/vimeo/psalm) in all packages.

Implementing psalm is quite easy.

#### Required
- [ ] Create a `.psalm.xml.dist` in the project root
- [ ] Copy and paste the contents from this [psalm.xml.dist](https://github.com/mezzio/repo-template/blob/6124090610ef1c6d6b0c80a063cafacc3ec68f38/psalm.xml.dist)
- [ ] Run `$ composer require vimeo/psalm`
- [ ] Run `$ vendor/bin/psalm --set-baseline=psalm-baseline.xml`
- [ ] Add a composer script `static-analysis` with the command `psalm --shepherd --stats`
- [ ] Add a new line to `script:` in `.travis.yml`: `- if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi`
- [ ] Remove phpstan from the project (`phpstan.neon.dist`, `.travis.yml` entry, `composer.json` `require-dev` and `scripts`)

##### Optional
- [ ] Fix as many psalm errors as possible.

### PR DESCRIPTION
**Pull request by ocean**
Added Psalm integration, as per #26 

Had to do a couple of extra steps from the suggestions in the issue, and managed to apply a bunch of automated fixes.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -26,7 +26,8 @@
         "ext-libxml": "*",
         "laminas/laminas-escaper": "^2.5.2",
         "laminas/laminas-stdlib": "^3.2.1",
-        "laminas/laminas-zendframework-bridge": "^1.0"
+        "laminas/laminas-zendframework-bridge": "^1.0",
+        "vimeo/psalm": "^3.16"
```
Comment (boesing) [inline composer.json]: Please move `vimeo/psalm` dependency to `require-dev`

File: src/PubSubHubbub/PubSubHubbub.php [hunk 2]
```diff
@@ -108,8 +108,10 @@ public static function clearHttpClient()
      * Set the Escaper instance
      *
      * If null, resets the instance
+     *
+     * @return void
      */
-    public static function setEscaper(Escaper $escaper = null)
+    public static function setEscaper(Escaper $escaper = null): void
```
Comment (boesing) [inline src/PubSubHubbub/PubSubHubbub.php]: Please do not add php typehints as this will is a BC break unless the class is `final`.

File: src/Reader/AbstractFeed.php [hunk 3]
```diff
@@ -91,8 +91,10 @@ public function __construct(DOMDocument $domDocument, $type = null)
      * a self-referencing URI.
      *
      * @param string $uri
+     *
+     * @return void
      */
-    public function setOriginalSourceUri($uri)
+    public function setOriginalSourceUri($uri): void
```
Comment (boesing) [inline src/Reader/AbstractFeed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/AbstractFeed.php [hunk 4]
```diff
@@ -237,7 +239,7 @@ public function valid()
         return 0 <= $this->entriesKey && $this->entriesKey < $this->count();
     }
 
-    public function getExtensions()
+    public function getExtensions(): array
```
Comment (boesing) [inline src/Reader/AbstractFeed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/AbstractFeed.php [hunk 5]
```diff
@@ -268,7 +270,7 @@ public function getExtension($name)
         return;
     }
 
-    protected function loadExtensions()
+    protected function loadExtensions(): void
```
Comment (boesing) [inline src/Reader/AbstractFeed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/Extension/AbstractFeed.php [hunk 6]
```diff
@@ -133,9 +133,9 @@ public function setXpath(DOMXPath $xpath = null)
     /**
      * Get the DOMXPath object
      *
-     * @return string
+     * @return DOMXPath
      */
-    public function getXpath()
+    public function getXpath(): DOMXPath
```
Comment (boesing) [inline src/Reader/Extension/AbstractFeed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/Extension/Atom/Entry.php [hunk 7]
```diff
@@ -545,10 +545,11 @@ public function getSource()
...
      *  xml:base value as a prefix to turn into an absolute URI.
      *
-     * @param  string $link
-     * @return string
+     * @param string $link
+     *
+     * @return null|string
      */
-    protected function absolutiseUri($link)
+    protected function absolutiseUri($link): ?string
```
Comment (boesing) [inline src/Reader/Extension/Atom/Entry.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/Extension/Atom/Entry.php [hunk 8]
```diff
@@ -564,7 +565,9 @@ protected function absolutiseUri($link)
     /**
      * Get an author entry
      *
-     * @return string
+     * @return (null|string)[]|null
+     *
+     * @psalm-return array{email?: null|string, name?: null|string, uri?: null|string}|null
```
Comment (boesing) [inline src/Reader/Extension/Atom/Entry.php:569]: AWESOME addition!

File: src/Reader/Extension/Atom/Entry.php [hunk 9]
```diff
@@ -610,9 +613,9 @@ protected function registerNamespaces()
     /**
      * Detect the presence of any Atom namespaces in use
      *
-     * @return string
+     * @return null|string
      */
-    protected function getAtomType()
+    protected function getAtomType(): ?string
```
Comment (boesing) [inline src/Reader/Extension/Atom/Entry.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/Extension/Atom/Feed.php [hunk 10]
```diff
@@ -445,7 +445,9 @@ public function getCategories()
     /**
      * Get an author entry in RSS format
      *
-     * @return string
+     * @return (null|string)[]|null
```
Comment (boesing) [inline src/Reader/Extension/Atom/Feed.php]: This should be annotated as `array<string,null|string>` as it is a dictionary.

File: src/Reader/Extension/Atom/Entry.php [hunk 11]
```diff
@@ -564,7 +565,9 @@ protected function absolutiseUri($link)
     /**
      * Get an author entry
      *
-     * @return string
+     * @return (null|string)[]|null
```
Comment (boesing) [inline src/Reader/Extension/Atom/Entry.php]: This should be annotated as array<string,null|string> as it is a dictionary.

File: src/Reader/Extension/Thread/Entry.php [hunk 12]
```diff
@@ -15,9 +15,9 @@ class Entry extends Extension\AbstractEntry
     /**
      * Get the "in-reply-to" value
      *
-     * @return string
+     * @return void
      */
-    public function getInReplyTo()
+    public function getInReplyTo(): void
```
Comment (boesing) [inline src/Reader/Extension/Thread/Entry.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/Feed/AbstractFeed.php [hunk 13]
```diff
@@ -90,8 +90,10 @@ public function __construct(DOMDocument $domDocument, $type = null)
      * a self-referencing URI.
      *
      * @param string $uri
+     *
+     * @return void
      */
-    public function setOriginalSourceUri($uri)
+    public function setOriginalSourceUri($uri): void
```
Comment (boesing) [inline src/Reader/Feed/AbstractFeed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/Feed/AbstractFeed.php [hunk 14]
```diff
@@ -236,7 +238,7 @@ public function valid()
         return 0 <= $this->entriesKey && $this->entriesKey < $this->count();
     }
 
-    public function getExtensions()
+    public function getExtensions(): array
```
Comment (boesing) [inline src/Reader/Feed/AbstractFeed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/Feed/AbstractFeed.php [hunk 15]
```diff
@@ -268,7 +270,7 @@ public function getExtension($name)
         return null;
     }
 
-    protected function loadExtensions()
+    protected function loadExtensions(): void
```
Comment (boesing) [inline src/Reader/Feed/AbstractFeed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/Feed/Atom.php [hunk 16]
```diff
@@ -143,9 +143,9 @@ public function getDateModified()
     /**
      * Get the feed lastBuild date. This is not implemented in Atom.
      *
-     * @return null|string
+     * @return void
      */
-    public function getLastBuildDate()
+    public function getLastBuildDate(): void
```
Comment (boesing) [inline src/Reader/Feed/Atom.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/Reader.php [hunk 17]
```diff
@@ -695,7 +699,7 @@ protected static function registerCoreExtensions()
      * @param  array
      * @return array
      */
-    public static function arrayUnique(array $array)
+    public static function arrayUnique(array $array): array
```
Comment (boesing) [inline src/Reader/Reader.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/Reader.php [hunk 18]
```diff
@@ -546,8 +548,10 @@ public static function detectType($feed, $specOnly = false)
 
     /**
      * Set plugin manager for use with Extensions
+     *
+     * @return void
      */
-    public static function setExtensionManager(ExtensionManagerInterface $extensionManager)
+    public static function setExtensionManager(ExtensionManagerInterface $extensionManager): void
```
Comment (boesing) [inline src/Reader/Reader.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/StandaloneExtensionManager.php [hunk 19]
```diff
@@ -58,8 +58,10 @@ public function get($extension)
      *
      * @param string $name
      * @param string $class
+     *
+     * @return void
      */
-    public function add($name, $class)
+    public function add($name, $class): void
```
Comment (boesing) [inline src/Reader/StandaloneExtensionManager.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Reader/StandaloneExtensionManager.php [hunk 20]
```diff
@@ -81,8 +83,10 @@ public function add($name, $class)
      * Remove an extension.
      *
      * @param string $name
+     *
+     * @return void
      */
-    public function remove($name)
+    public function remove($name): void
```
Comment (boesing) [inline src/Reader/StandaloneExtensionManager.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/AbstractFeed.php [hunk 21]
```diff
@@ -126,11 +126,14 @@ public function setCopyright($copyright)
...
-     * @return $this
+     * @param null|int|DateTimeInterface
+     * @param DateTime|\DateTimeImmutable|int|null|string $date
+     *
+     * @return self
+     *
      * @throws Exception\InvalidArgumentException
      */
-    public function setDateCreated($date = null)
+    public function setDateCreated($date = null): self
```
Comment (boesing) [inline src/Writer/AbstractFeed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/AbstractFeed.php [hunk 22]
```diff
@@ -151,11 +154,14 @@ public function setDateCreated($date = null)
...
-     * @return $this
+     * @param null|int|DateTimeInterface
+     * @param DateTime|\DateTimeImmutable|int|null|string $date
+     *
+     * @return self
+     *
      * @throws Exception\InvalidArgumentException
      */
-    public function setDateModified($date = null)
+    public function setDateModified($date = null): self
```
Comment (boesing) [inline src/Writer/AbstractFeed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/AbstractFeed.php [hunk 23]
```diff
@@ -176,11 +182,14 @@ public function setDateModified($date = null)
...
-     * @return $this
+     * @param null|int|DateTimeInterface
+     * @param DateTime|\DateTimeImmutable|int|null|string $date
+     *
+     * @return self
+     *
      * @throws Exception\InvalidArgumentException
      */
-    public function setLastBuildDate($date = null)
+    public function setLastBuildDate($date = null): self
```
Comment (boesing) [inline src/Writer/AbstractFeed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/Deleted.php [hunk 24]
```diff
@@ -32,12 +32,13 @@ class Deleted
...
-     * @return null|string
-     * @return Deleted
+     * @param null|string $encoding
+     *
+     * @return self
+     *
      * @throws Exception\InvalidArgumentException
      */
-    public function setEncoding($encoding)
+    public function setEncoding(?string $encoding): self
```
Comment (boesing) [inline src/Writer/Deleted.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/Extension/GooglePlayPodcast/Entry.php [hunk 25]
```diff
@@ -66,10 +66,13 @@ public function getEncoding()
...
      *
-     * @param  string
+     * @param string
+     *
      * @throws Writer\Exception\InvalidArgumentException
+     *
+     * @return void
      */
-    public function setPlayPodcastBlock($value)
+    public function setPlayPodcastBlock($value): void
```
Comment (boesing) [inline src/Writer/Extension/GooglePlayPodcast/Entry.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/Extension/GooglePlayPodcast/Feed.php [hunk 26]
```diff
@@ -67,11 +67,13 @@ public function getEncoding()
...
-     * @param  string
-     * @return $this
+     * @param string
+     *
+     * @return self
+     *
      * @throws Writer\Exception\InvalidArgumentException
      */
-    public function setPlayPodcastBlock($value)
+    public function setPlayPodcastBlock($value): self
```
Comment (boesing) [inline src/Writer/Extension/GooglePlayPodcast/Feed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/Extension/ITunes/Entry.php [hunk 27]
```diff
@@ -67,10 +67,13 @@ public function getEncoding()
...
      *
-     * @param  string
+     * @param string
+     *
      * @throws Writer\Exception\InvalidArgumentException
+     *
+     * @return void
      */
-    public function setItunesBlock($value)
+    public function setItunesBlock($value): void
```
Comment (boesing) [inline src/Writer/Extension/ITunes/Entry.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/Extension/ITunes/Feed.php [hunk 28]
```diff
@@ -67,11 +67,13 @@ public function getEncoding()
...
-     * @param  string
-     * @return $this
+     * @param string
+     *
+     * @return self
+     *
      * @throws Writer\Exception\InvalidArgumentException
      */
-    public function setItunesBlock($value)
+    public function setItunesBlock($value): self
```
Comment (boesing) [inline src/Writer/Extension/ITunes/Feed.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/Renderer/AbstractRenderer.php [hunk 29]
```diff
@@ -167,8 +167,10 @@ public function getExceptions()
      * on their appropriateness for the current type, e.g. renderers.
      *
      * @param string $type
+     *
+     * @return void
      */
-    public function setType($type)
+    public function setType($type): void
```
Comment (boesing) [inline src/Writer/Renderer/AbstractRenderer.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/Renderer/AbstractRenderer.php [hunk 30]
```diff
@@ -188,8 +190,10 @@ public function getType()
      * helps simplify the appending of namespace declarations, but also ensures
      * namespaces are added to the root element - not scattered across the entire
      * XML file - may assist namespace unsafe parsers and looks pretty ;).
+     *
+     * @return void
      */
-    public function setRootElement(DOMElement $root)
+    public function setRootElement(DOMElement $root): void
```
Comment (boesing) [inline src/Writer/Renderer/AbstractRenderer.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/StandaloneExtensionManager.php [hunk 31]
```diff
@@ -58,8 +58,10 @@ public function get($extension)
      *
      * @param string $name
      * @param string $class
+     *
+     * @return void
      */
-    public function add($name, $class)
+    public function add($name, $class): void
```
Comment (boesing) [inline src/Writer/StandaloneExtensionManager.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/StandaloneExtensionManager.php [hunk 32]
```diff
@@ -83,8 +85,10 @@ public function add($name, $class)
      * Remove an extension.
      *
      * @param string $name
+     *
+     * @return void
      */
-    public function remove($name)
+    public function remove($name): void
```
Comment (boesing) [inline src/Writer/StandaloneExtensionManager.php]: Please do not add php typehints as this will is a BC break unless the class is final.

File: src/Writer/Writer.php [hunk 33]
```diff
@@ -60,8 +60,10 @@ class Writer
      * Set plugin loader for use with Extensions
      *
      * @param ExtensionManagerInterface
+     *
+     * @return void
      */
-    public static function setExtensionManager(ExtensionManagerInterface $extensionManager)
+    public static function setExtensionManager(ExtensionManagerInterface $extensionManager): void
```
Comment (boesing) [inline src/Writer/Writer.php]: Please do not add php typehints as this will is a BC break unless the class is final.

**Review by boesing — CHANGES_REQUESTED** [review]

File: src/Reader/Extension/Atom/Entry.php (hunk 8 above)
Comment (ocean) [reply src/Reader/Extension/Atom/Entry.php:569]: Thanks, but this was Psalm's doing, not me directly 😁  Obviously a good decision to adopt this tool 👍

File: src/PubSubHubbub/PubSubHubbub.php (hunk 2 above)
Comment (ocean) [reply src/PubSubHubbub/PubSubHubbub.php]: Ok to have the comment @ return type annotations though?

Comment (boesing) [reply src/PubSubHubbub/PubSubHubbub.php]: Yes, `@return` is fine. Sadly, the native type hint is a problem in case there is a class which is overwriting this method by using inheritance....

File: src/Reader/Feed/AbstractFeed.php (hunk 14 above)
Comment (boesing) [reply src/Reader/Feed/AbstractFeed.php]: probably you missed this one 🙈

File: src/Reader/Feed/FeedInterface.php [hunk 34]
```diff
@@ -51,6 +51,13 @@ public function getDateCreated();
      */
     public function getDateModified();
 
+    /**
+     * Get the feed lastBuild date
+     *
+     * @return null|DateTime
+     */
+    public function getLastBuildDate();
```
Comment (boesing) [inline src/Reader/Feed/FeedInterface.php]: Please only focus on adding psalm instead of adding features like this.
We should not add new methods to interfaces. If there are issues with psalm, just add these to the baseline.

Would love to see this as a separated PR. *Maybe* we will add it to the next release tho.

File: src/Reader/Feed/Rss.php [hunk 35]
```diff
@@ -474,6 +474,16 @@ public function getLanguage()
         return $this->data['language'];
     }
 
+    /**
+     * Get a link to the source website. This is not implemented in RSS.
+     *
+     * @return void
+     */
+    public function getBaseUrl()
```
Comment (boesing) [inline src/Reader/Feed/Rss.php]: Please only focus on adding psalm instead of adding features like this.
We should not add new methods to interfaces. If there are issues with psalm, just add these to the baseline.

Would love to see this as a separated PR. Maybe we will add it to the next release tho.

File: src/Writer/Deleted.php [hunk 36]
```diff
@@ -32,14 +32,15 @@ class Deleted
...
-     * @return null|string
-     * @return Deleted
+     * @param null|string $encoding
+     *
+     * @return self
+     *
      * @throws Exception\InvalidArgumentException
      */
-    public function setEncoding($encoding)
+    public function setEncoding(?string $encoding)
```
Comment (boesing) [inline src/Writer/Deleted.php]: Probably this one slipped through aswell.

File: src/Writer/Deleted.php [hunk 37]
```diff
@@ -32,14 +32,15 @@ class Deleted
     /**
      * Set the feed character encoding
      *
-     * @param  string $encoding
-     * @return null|string
-     * @return Deleted
+     * @param null|string $encoding
```
Comment (boesing) [inline src/Writer/Deleted.php]: Imho we should not allow `null` at all as it was not previously declared like that (or is that a conflict with the implementing interface?).

Even if there is code which passes `null` to this method in some cases, we should let the `!is_string` do its job.
Please revert this change and add errors regarding to the old annotation to the baseline.

File: src/Writer/Deleted.php [hunk 38]
```diff
@@ -32,14 +32,15 @@ class Deleted
...
+     *
+     * @return self
+     *
      * @throws Exception\InvalidArgumentException
      */
-    public function setEncoding($encoding)
+    public function setEncoding(?string $encoding)
     {
-        if (empty($encoding) || ! is_string($encoding)) {
+        if (empty($encoding)) {
```
Comment (boesing) [inline src/Writer/Deleted.php]: see comment on line 35. please do not change code behavior.

**Review by boesing — CHANGES_REQUESTED** [review]

**Comment by ocean** [comment]
@boesing Ok, will concentrate on the Psalm integration instead of making any class changes 👍 

Thanks for your patient comments, encouragement and guidance through this, hopefully I'll get it done today 😃

File: src/Writer/Deleted.php (hunk 37 above)
Comment (ocean) [reply src/Writer/Deleted.php]: Have reverted this to how it was - it has 2 return annotations, which Psalm really doesn't like, heh. I'll fix in another PR.

**Comment by ocean** [comment]
TravisCI test failures seem to be caused by an error in the test code of a dependency, `laminas-servicemanager`, in `/vendor/laminas/laminas-servicemanager/src/Test/CommonPluginManagerTrait.php`.

File: src/Reader/AbstractFeed.php [hunk 39]
```diff
@@ -91,6 +91,8 @@ public function __construct(DOMDocument $domDocument, $type = null)
      * a self-referencing URI.
      *
      * @param string $uri
+     *
+     * @return void
```
Comment (froschdesign) [inline src/Reader/AbstractFeed.php]: The empty line is not allowed. Please compare with the coding standard: "[9. Commenting and DocBlocks](https://docs.laminas.dev/laminas-coding-standard/v2/coding-style-guide/#9-commenting-and-docblocks)"

Thanks in advance!

Comment (ocean) [reply src/Reader/AbstractFeed.php]: Ok, I found a few of those with this regex I put together `\*\s@param\s(\w+\s\$\w+)$\n\s+\*$`, and fixed them.

**Review by boesing — APPROVED** [review]
LGTM. 

As already stated in your other PR, we have to wait for PHP 8.0 which drops PHP <7.3 aswell and upgrades phpunit.
`TestCase::setExpectedException` was removed somewhen in the last phpunit versions and was replaced by `TestCase::assertException`.

**Comment by boesing** [comment]
@ocean I've updated your branch. tests should now pass but I screwed up with the DCO. Please execute the commands mentioned here: https://github.com/laminas/laminas-feed/pull/28/checks?check_run_id=1336866873

After this, we can merge this and are ready for release.

**Comment by froschdesign** [comment]
@ocean 
We need your help here! A release is waiting and we want to add your work. [Your commits needs a correct sign off.](https://github.com/laminas/laminas-feed/pull/28/checks?check_run_id=1336868041)

Thanks in advance! 👍

**Comment by ocean** [comment]
Sorry guys, been a bit of a busy week. Coming right up.

**Comment by ocean** [comment]
Thanks for the additions and fixes @boesing and for the prompting @froschdesign 😃 

Fixed the DCO issue but now the latest Travis run has failed with Psalm errors.

https://travis-ci.com/github/laminas/laminas-feed/jobs/429737672#L735

**Comment by ocean** [comment]
@boesing Should I just update the Psalm baseline to ignore these errors, given that the tests are passing?

( https://travis-ci.com/github/laminas/laminas-feed/jobs/429737672#L735 )

**Comment by boesing** [comment]
@ocean Sorry for being late. Had plenty of stuff to do.
Thanks for fixing this up.
Yes, feel free to add stuff to the baseline unless they're not added by this PR.

I'll merge and release this package this week, if we can get this merged 👍

**Comment by ocean** [comment]
🎉

**Comment by boesing** [comment]
Hm, I probably f*** up with the DCO again. Not sure why my git is always signing up aswell.
@ocean Could you fix DCO one more time please? 😅

**Comment by weierophinney** [comment]
> Hm, I probably f*** up with the DCO again. Not sure why my git is always signing up aswell.
> @ocean Could you fix DCO one more time please?

DCO checks are green. :wink:
